torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -1,269 +1,266 @@
+    max_val = _convert_to_int(range.max_val)
+    return min_val, max_val
+
+
 class _AddRuntimeAssertionsForConstraintsPass(ExportPassBase):
     def __init__(
         self,
-        input_shape_constraints,
-        input_name_to_example_inputs,
-        inline_constraints,
-    ) -> None:
+        symbol_to_range: Dict[sympy.Symbol, RangeConstraint],
+        equality_constraints: Dict[NodeDim, List[NodeDim]],
+    ):
         super().__init__()
-        self.constraints = self._process_shape_constraints(input_shape_constraints)
-        self.input_name_to_example_inputs = input_name_to_example_inputs
-        self.inline_constraints = inline_constraints
-        self.input_name_to_args: Dict[str, ProxyValue] = {}
-
-    def _process_shape_constraints(self, constraints) -> Dict[str, List[ConstraintSpec]]:
-        input_name_to_dim_constraints: Dict[str, ConstraintsContainer] = defaultdict(
-            lambda: ConstraintsContainer(defaultdict(list), defaultdict(list))
+        self.symbol_to_range: Dict[sympy.Symbol, RangeConstraint] = symbol_to_range
+        self.equality_constraints: Dict[NodeDim, List[NodeDim]] = equality_constraints
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        graph_module = copy.deepcopy(graph_module)
+        graph = graph_module.graph
+
+        # Add runtime asserts for input shape constraints. We do this after all
+        # placeholder nodes so that we can handle both (unary) predicates and
+        # (binary) relations.
+        placeholder_nodes: Dict[str, torch.fx.Node] = OrderedDict()
+        for node in graph.nodes:
+            if node.op != "placeholder":
+                continue
+
+            placeholder_nodes[node.name] = node
+            if (
+                "val" not in node.meta or
+                not isinstance(node.meta["val"], FakeTensor)
+            ):
+                continue
+
+            fake_tensor_shape = node.meta["val"].shape
+            for dim, shape in enumerate(fake_tensor_shape):
+                if isinstance(shape, SymInt):
+                    # If the shape is dynamic, add range assertions
+                    symbol = shape.node._expr
+                    assert symbol in self.symbol_to_range
+
+                    with graph.inserting_after(node):
+                        self._insert_range_assert_inplace(
+                            graph, node, dim, self.symbol_to_range[symbol]
+                        )
+                else:
+                    # If no dynamism is specified, we assume all dimensions #
+                    # are specialized
+                    assert isinstance(shape, int)
+                    with graph.inserting_after(node):
+                        self._insert_specialized_shape_assert_inplace(
+                            graph, node, dim, shape,
+                        )
+
+        # Add runtime assertions on equality constraints on the inputs
+        with graph.inserting_after(
```

Comment:
Ah, TIL. So you're inserting right after the last placeholder, is that correct?

Not here, but it would be cool if we could do everything here (including unary) and create only one new node per `NodeDim` so that we could reuse them across all constraints.

Reply:
Yup, I'm inserting after the last placeholder. I chose to do it this ahead of running the interpreter based pass because the previous method would've inserted input constraints on every subgraph, but we only care about the initial subgraph's placeholder nodes.

> Not here, but it would be cool if we could do everything here (including unary) and create only one new node per NodeDim so that we could reuse them across all constraints.

Yup, this makes sense!